enhancements/microshift/low-latency-workloads-on-microshift.md

```diff
@@ -0,0 +1,737 @@
+
+
+### API Extensions
+
+Following API extensions are expected:
+- A passthrough from MicroShift's config to Kubelet config.
+- Variables file for TuneD profile to allow customization of the profile for specific host.
+
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+N/A
+
+#### Standalone Clusters
+
+N/A
+
+#### Single-node Deployments or MicroShift
+
+Purely MicroShift enhancement.
+
+### Implementation Details/Notes/Constraints
+
+#### TuneD Profile
+
+New `microshift-baseline` tuned profile will be created and will include existing `cpu-partitioning` profile.
+
+`/etc/tuned/microshift-baseline-variables.conf` will be used by users to provide custom values for settings such as:
+- isolated CPU set
+- hugepage count (both 2M and 1G)
+- additional kernel arguments
+
+```ini
+[main]
+summary=Optimize for running low latency workloads on MicroShift
+include=cpu-partitioning
+
+[variables]
+include=/etc/tuned/microshift-baseline-variables.conf
+
+[bootloader]
+cmdline_microshift=+default_hugepagesz=${hugepagesDefaultSize} hugepagesz=2M hugepages=${hugepages2M} hugepagesz=1G hugepages=${hugepages1G}
+cmdline_additionalArgs=+${additionalArgs}
+```
+
+```ini
+### cpu-partitioning variables
+#
+# Core isolation
+#
+# The 'isolated_cores=' variable below controls which cores should be
+# isolated. By default we reserve 1 core per socket for housekeeping
+# and isolate the rest. But you can isolate any range as shown in the
+# examples below. Just remember to keep only one isolated_cores= line.
+#
+# Examples:
+# isolated_cores=2,4-7
+# isolated_cores=2-23
+#
+# Reserve 1 core per socket for housekeeping, isolate the rest.
+# Change this for a core list or range as shown above.
+isolated_cores=${f:calc_isolated_cores:1}
+
+# To disable the kernel load balancing in certain isolated CPUs:
+# no_balance_cores=5-10
+
+### microshift-baseline variables
+# Default hugepages size
+hugepagesDefaultSize = 2M
+
+# Amount of 2M hugepages
+hugepages2M = 128
+
+# Amount of 1G hugepages
+hugepages1G = 0
+
+# Additional kernel arguments
+additionalArgs = ""
+```
+
+#### `microshift-tuned.service` configuration
+
+Config file to specify which profile to re-apply each boot and if host should be rebooted if
+the kargs before and after applying profile are mismatched.
+
+```yaml
+profile: microshift-baseline
+reboot_after_apply: True
+```
+
+#### CRI-O configuration
+
+```ini
+[crio.runtime.runtimes.high-performance]
+runtime_path = "/bin/crun"
+runtime_type = "oci"
+runtime_root = "/bin/crun"
+allowed_annotations = ["cpu-load-balancing.crio.io", "cpu-quota.crio.io", "irq-load-balancing.crio.io", "cpu-c-states.crio.io", "cpu-freq-governor.crio.io"]
+```
+
+#### Kubelet configuration 
+
+Because of multitude of option in Kubelet configuration, a simple passthrough (copy paste)
+will be implemented, rather than exposing every single little configuration variable.
+
+```yaml
+# /etc/microshift/config.yaml
+kubelet:
```

Comment:
Are these defaults something the user should not change? Or is it all subject to manual configuration?

Reply:
These are example values I borrowed from @sjug's "Low Latency on MicroShift" documents.
There are some that we expect to be specific value (like cpu/memory managers "static" policy, etc.) so that other parts work as expected for low latency/workload partitioning. Other, numeric values are host/setup/intention dependent, so this is something users can tweak.

But runtime specific, we don't apply any of these by default. They need to be explicitly provided in the config file to make them work.